### ISSUE TITLE
Instantiating a large Jacobian at every loop

### ISSUE THREAD

**Issue by hanrach** [google/jax#4753]
I am trying to use the jax package to build a Jacobian to solve for nonlinear equations via Newton's method. See [here](https://github.com/hanrach/p2d_solver/blob/main/run_ex.py) for the code. I explicitly instantiate Jacobian of size ~5000X5000 at every Newton iteration, and I'm looking for a faster way to speed up the computation. The program uses about 3640+MB of memory (197% of my CPU) and 5 Newton iterations takes around 252 seconds. 

I create jacobian functions outside the loop,
```
jac_fn = jax.jit(jacfwd(fn))
fn = jax.jit(fn)
```

where `fn` takes a vector of size ~5000. Inside newton's method,
```
 while(count < maxit and  res > tol):
        start1 =timeit.default_timer() 
        J =  jac_fn(U, Uold)
        y = fn(U,Uold)
        res = norm(y/norm(U,np.inf),np.inf)
        delta = solve(J,y)
        U = U - delta
        count = count + 1
        end1 =timeit.default_timer() 
        print(count, res)
```

My biggest concern is that the large matrix at every iteration is a dense matrix so it takes up a lot of memory (and possibly the linear solve is slower than the sparse version?) I believe that Jax doesn't have implementations for something like scipy.sparse. Could this be the issue? When I print out the time taken in each loop, they're less than a second except the first loop. I don't quite understand how this can add up to couple of minutes.

I'm using `jacfwd` instead of `jacrev` as  the process is automatically killed with `jacrev` ( I think because of memory issues) and my function does not have more arguments than outputs.

Issue #2302 has a similar problem I believe.

**Comment by C-J-Cundy** [comment]
Something you could consider doing is using the [JAX conjugate gradients solver](https://jax.readthedocs.io/en/latest/_autosummary/jax.scipy.sparse.linalg.cg.html). 

The idea is that you can solve a system Ax = b without needing to instantiate A, you just need a function that gives x computes Ax. You want to solve the system J @ delta = y for Newton's method. 
The function 
```
jac_x_prod = lambda x: jvp(fn, y, x)
```
is exactly this function that compute J @ x given x, where J is the Jacobian of `fn` evaluated at y.
I think you might be able to use `linearize` here too, but I'm not too familiar with that.

So in summary, if you replace `delta = solve(J,y)` with 

```
jac_x_prod = lambda x: jvp(fn, y, x)
delta = jax.scipy.sparse.linalg.cg(jax_x_prod, y)[0]
```
This should drastically cut down your memory usage. 
As for whether it would be in general faster, I'm not too sure.

**Comment by froystig** [comment]
Indeed, `jax.scipy.sparse.linalg.cg` is a matrix-free solver that we have available today. Any matrix-free solver would allow you to use `jax.jvp` ("Jacobian-vector product") rather than necessarily instantiating a full Jacobian matrix.

I'm going to close this issue since "instantiating a large Jacobian" should be addressed. Feel free to re-open if you continue to encounter something more.

In case CG is not the right algorithm for this application somehow, we'd welcome contributions of other matrix-free solvers in `scipy.linalg`!
